pytorch_lightning/utilities/auto_restart.py

```diff
@@ -570,4 +598,146 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
         )
 
     else:
-        raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+        if _fault_tolerant_training_mode().is_automatic:
+            raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+        latest_worker_id = state_dict["latest_worker_id"]
+        num_workers = state_dict["state"][latest_worker_id]["num_workers"]
+        sampler_state = state_dict["state"][latest_worker_id]["sampler_state"]
+        if sampler_state:
+            for k in sampler_state:
+                obj = getattr(dataloader, k)
+                if not is_obj_stateful(obj):
+                    raise MisconfigurationException(
+                        f"The DataLoader attribute should have a `load_state_dict` method. Found {obj}"
+                    )
+
+                obj.load_state_dict(sampler_state[k])
+
+        if not hasattr(dataset, "load_state_dict"):
+            return
+
+        dataset_state = {
+            worker_id: state_dict["state"][worker_id]["dataset_state"][worker_id]
+            for worker_id in state_dict["state"].keys()
+        }
+
+        if num_workers > 0:
+            # remap states to worker ids starting at 0
+            next_worker_id = latest_worker_id + 1
+            old_to_new_worker_id_map = [((next_worker_id + i) % num_workers, i) for i in range(num_workers)]
+            dataset_state = {
+                new_id: dataset_state[old_id] for old_id, new_id in old_to_new_worker_id_map if old_id in dataset_state
+            }
```

Comment:
this code is duplicated from CaptureMapDataset. It already gets done in load_state_dict() of CaptureMapDataset which is what gets called in the line below. 
So there is probably a bug here because it will rotate the states twice.

Reply:
Fault manual, there is no CaptureDataset been used, meaning we need to rotate the indices. I extracted this logic into its own function.

Reply:
Done !